sedaro/src/sedaro/branches/scenario_branch/sim_client.py

```diff
@@ -773,3 +776,37 @@ def produce(self, agent_id: str, external_state_id: str, values: tuple, timestam
                 **COMMON_API_KWARGS,
             )
         return tuple(serdes(v) for v in body_from_res(response))
+
+    @asynccontextmanager
+    async def async_channel(self, grpc_host: str):
+        """
+        Asynchronous context manager for cosimulation sessions.
+        Automatically opens a cosimulation grpc session on entry and closes it on exit.
+
+        Usage:
+            async with simulation_handle.async_channel(grpc_host="...") as cosim_client:
+                # ...
+        """
+        sedaro = self._sim_client.get_sedaro()
+        api_key = sedaro._api_key
+        host = sedaro._api_host
+
+        status = self._sim_client.status()
+        address = status.get("clusterAddr")
+        job_id = status.get("id")
+
+        cosim_client = CosimClient(
```

Comment:
The API key is used here to get auth JWTs from django to use with the cosim service.  Then the cosim service returns a session token so that we can refresh JWTs but maintain a session.

Reply:
I see - so if I'm understanding correctly, we need to add plumbing here for auth_handles?

Reply:
Yes, you're right.  Now that I look at it, I think `_get_auth_token` should be reaching outside of the async client and using the api client to do this request.

Reply:
I created an issue for this in gitlab.